File: sdk/encryption/encryption.go
package encryption

import (
	"crypto/rand"
	"encoding/binary"
	"errors"

	"golang.org/x/crypto/chacha20poly1305"

	"github.com/bpprotocol/blockparty/sdk/blockpb"
	"github.com/bpprotocol/blockparty/sdk/crypto"
)

// NonceSize is the XChaCha20-Poly1305 nonce length prefixed to the data field.
const NonceSize = chacha20poly1305.NonceSizeX // 24

// aeadInfo is the HKDF info prefix for content-key derivation; the block nonce
// is appended to it.
const aeadInfo = "bpprotocol.org/v1/aead"

var (
	// ErrShortData is returned when a data field is too short to contain a nonce.
	ErrShortData = errors.New("encryption: data shorter than nonce")
	// ErrDecrypt is returned when authentication fails: wrong audience secret,
	// tampered ciphertext, or tampered AAD-bound metadata.
	ErrDecrypt = errors.New("encryption: authentication failed")
)

// ContentKey derives the 32-byte per-block content key from the audience
// secret, the audience code, and the block nonce (see encryption.md).
func ContentKey(audienceSecret, audienceCode, blockNonce []byte) []byte {
	info := append([]byte(aeadInfo), blockNonce...)
	return crypto.HKDFSHA256(audienceSecret, audienceCode, info, 32)
}

// AAD builds the canonical additional authenticated data bound to a block's
// ciphertext: the 4-byte big-endian version, the raw type and audience codes,
// then the 8-byte big-endian timestamp.
func AAD(version uint32, typeCode, audienceCode []byte, timestamp int64) []byte {
	out := make([]byte, 0, 4+len(typeCode)+len(audienceCode)+8)
	out = binary.BigEndian.AppendUint32(out, version)
	out = append(out, typeCode...)
	out = append(out, audienceCode...)
	out = binary.BigEndian.AppendUint64(out, uint64(timestamp))
	return out
}

// EncryptData encrypts plaintext for an audience and returns the data field:
// a fresh 24-byte nonce followed by the AEAD ciphertext (which includes the
// Poly1305 tag).
func EncryptData(audienceSecret, audienceCode, plaintext, aad []byte) ([]byte, error) {
	nonce := make([]byte, NonceSize)
	if _, err := rand.Read(nonce); err != nil {
		return nil, err
	}
	return sealWithNonce(audienceSecret, audienceCode, nonce, plaintext, aad)
}

// EncryptForBlock is a convenience that builds the AAD from a block's metadata
// and encrypts plaintext into a data field. The caller then constructs the
// block with the returned data (so the block ID binds the ciphertext).
func EncryptForBlock(audienceSecret []byte, version uint32, typeCode, audienceCode []byte, timestamp int64, plaintext []byte) ([]byte, error) {
	return EncryptData(audienceSecret, audienceCode, plaintext, AAD(version, typeCode, audienceCode, timestamp))
}

// sealWithNonce performs the AEAD seal with a caller-supplied nonce. It is the
// deterministic core of EncryptData, used directly only by tests/vectors.
func sealWithNonce(audienceSecret, audienceCode, nonce, plaintext, aad []byte) ([]byte, error) {
	aead, err := chacha20poly1305.NewX(ContentKey(audienceSecret, audienceCode, nonce))
	if err != nil {
		return nil, err
	}
	// Prefix the nonce, then append the sealed ciphertext after it.
	out := make([]byte, NonceSize, NonceSize+len(plaintext)+aead.Overhead())
	copy(out, nonce)
	return aead.Seal(out, nonce, plaintext, aad), nil
}

// DecryptData decrypts a block's data field using the audience secret. It
// reconstructs the AAD from the block's metadata, so any tampering with
// version, type_code, audience_code, or timestamp causes authentication to
// fail. A caller who cannot derive the audience secret cannot decrypt, and the
// block remains an opaque stub.
func DecryptData(audienceSecret []byte, b *blockpb.Block) ([]byte, error) {
	if len(b.Data) < NonceSize {
		return nil, ErrShortData
	}
	nonce, ciphertext := b.Data[:NonceSize], b.Data[NonceSize:]
	aead, err := chacha20poly1305.NewX(ContentKey(audienceSecret, b.AudienceCode, nonce))
	if err != nil {
		return nil, err
	}
	aad := AAD(b.Version, b.TypeCode, b.AudienceCode, b.Timestamp)
	plaintext, err := aead.Open(nil, nonce, ciphertext, aad)
	if err != nil {
		return nil, ErrDecrypt
	}
	return plaintext, nil
}

File: js/packages/sdk/src/encryption/encryption.ts
import { xchacha20poly1305 } from "@noble/ciphers/chacha.js";
import { randomBytes } from "@noble/hashes/utils.js";
import { hkdfSha256 } from "../crypto/index.js";
import { concatBytes, utf8ToBytes } from "../bytes.js";
import type { Block } from "../blockpb/block_pb.js";

/** XChaCha20-Poly1305 nonce length, prefixed to the data field. */
export const NONCE_SIZE = 24;

// HKDF info prefix for content-key derivation; the block nonce is appended.
const AEAD_INFO = "bpprotocol.org/v1/aead";

/**
 * Derive the 32-byte per-block content key from the audience secret, audience
 * code, and block nonce (see protocol/specs/encryption.md).
 */
export function contentKey(
  audienceSecret: Uint8Array,
  audienceCode: Uint8Array,
  blockNonce: Uint8Array,
): Uint8Array {
  const info = concatBytes(utf8ToBytes(AEAD_INFO), blockNonce);
  return hkdfSha256(audienceSecret, audienceCode, info, 32);
}

/**
 * Build the additional authenticated data bound to a block's ciphertext: the
 * 4-byte big-endian version, the raw type and audience codes, then the 8-byte
 * big-endian timestamp.
 */
export function aad(
  version: number,
  typeCode: Uint8Array,
  audienceCode: Uint8Array,
  timestamp: number | bigint,
): Uint8Array {
  const head = new Uint8Array(4);
  new DataView(head.buffer).setUint32(0, version >>> 0, false);
  const tail = new Uint8Array(8);
  new DataView(tail.buffer).setBigUint64(0, BigInt.asUintN(64, BigInt(timestamp)), false);
  return concatBytes(head, typeCode, audienceCode, tail);
}

/**
 * Deterministic AEAD seal with a caller-supplied nonce: returns
 * `nonce || ciphertext+tag`. Exposed for conformance vectors and deterministic
 * testing; normal callers use encryptData / encryptForBlock.
 */
export function sealWithNonce(
  audienceSecret: Uint8Array,
  audienceCode: Uint8Array,
  nonce: Uint8Array,
  plaintext: Uint8Array,
  aadBytes: Uint8Array,
): Uint8Array {
  const key = contentKey(audienceSecret, audienceCode, nonce);
  const ciphertext = xchacha20poly1305(key, nonce, aadBytes).encrypt(plaintext);
  return concatBytes(nonce, ciphertext);
}

/**
 * Encrypt plaintext for an audience and return the data field: a fresh 24-byte
 * nonce followed by the AEAD ciphertext (which includes the Poly1305 tag).
 */
export function encryptData(
  audienceSecret: Uint8Array,
  audienceCode: Uint8Array,
  plaintext: Uint8Array,
  aadBytes: Uint8Array,
): Uint8Array {
  return sealWithNonce(audienceSecret, audienceCode, randomBytes(NONCE_SIZE), plaintext, aadBytes);
}

/**
 * Convenience that builds the AAD from a block's metadata and encrypts plaintext
 * into a data field. The caller then constructs the block with the returned data
 * (so the block ID binds the ciphertext).
 */
export function encryptForBlock(
  audienceSecret: Uint8Array,
  version: number,
  typeCode: Uint8Array,
  audienceCode: Uint8Array,
  timestamp: number | bigint,
  plaintext: Uint8Array,
): Uint8Array {
  return encryptData(
    audienceSecret,
    audienceCode,
    plaintext,
    aad(version, typeCode, audienceCode, timestamp),
  );
}

/**
 * Decrypt a block's data field using the audience secret. The AAD is
 * reconstructed from the block's metadata, so tampering with version, type_code,
 * audience_code, or timestamp causes authentication to fail (decrypt throws).
 * A caller who cannot derive the audience secret cannot decrypt.
 */
export function decryptData(audienceSecret: Uint8Array, block: Block): Uint8Array {
  if (block.data.length < NONCE_SIZE) {
    throw new Error("encryption: data shorter than nonce");
  }
  const nonce = block.data.subarray(0, NONCE_SIZE);
  const ciphertext = block.data.subarray(NONCE_SIZE);
  const key = contentKey(audienceSecret, block.audienceCode, nonce);
  const aadBytes = aad(block.version, block.typeCode, block.audienceCode, block.timestamp);
  return xchacha20poly1305(key, nonce, aadBytes).decrypt(ciphertext);
}
